text/0000-eliminate-finalize.md

```diff
@@ -0,0 +1,142 @@
+
+1. Provide the kernel and show that it was confirmed on-chain.
+2. Prove knowledge of `sender.nonce`
+3. Show that `sender.public_nonce + receiver.public_nonce + (proof_nonce * G)` = `total_nonce * G` (ie. the `k*G` in the kernel signature)
+4. Provide the preimage to `proof_nonce` (`diffie_secret | receiver.public_excess | tx.amount`)
+
+## Slate Format
+
+A slate is now only needed for passing information from the sender to receiver. It must include the following data:
+
+* `tx.amount`
+* `tx.offset`
+* `tx.kernel_features`
+* `tx.fee`
+* `tx.inputs`
+* `receiver.change_output`
+* `receiver.public_excess` & `receiver.public_nonce` (to prevent the need for grinding)
+* `sender.public_excess`, `sender.public_nonce`, & `sender.partial_sig`
+* `time_to_live` (optional)
+
+TODO: Define JSON & binary formats
+TODO: Mention QR codes
+
+## Backward Compatibility
+
+TODO: Decide how to handle versioning. In theory, this should be simple. If given an old slatepack address, use the old slate format. If given a new, one-time address, then the new slate format is supported.
+
+## Invoices
+
+To send an invoice to a payer, the recipient provides the payer with a newly generated one-time address, an amount, and any additional fields we decide to support (memo, refund address, expiration, etc). The payer, upon approving payment of the invoice, simply follows the standard 2-step send process described in this RFC. There's no need for a special workflow.
+
+## Security
+
+### Nonce Reuse
+
+Nonces should be randomly generated by the receiving wallet, rather than derived from seed, since reuse of a nonce could lead to loss of funds. Receiving wallets should keep track of all one-time addresses (excesses & nonces) they generate in a list (`address_list`), along with an indicator for whether they have been used yet.
+
+Upon receipt of a partial transaction (slate), the receiver should obtain a lock on `address_list` and then check whether the address has been used before. If it has not, it's safe to finalize the transaction, but before broadcasting, the address should be marked as used in the `address_list`, and then the lock can be safely released.
+
+### Play Attacks
+
+TODO: Describe play attacks, and finish this section.
+
+Any inputs sent from a wallet to a one-time address should be considered as sent. They should not be cancelable, however, the sender should have the option to add additional inputs and/or increase the fee (equivalent of bitcoin's RBF).
+
+Receiver should be unable to claim a problem with the transaction. If they do, it's trivial to show them the slate that was sent, and prove that it's a valid partial transaction containing everything necessary for the receiver to finalize.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Addresses are significantly longer (170 characters).
+
+Using one-time addresses increases the risk of "Play" attacks [2]. Recommendations for dealing with these have been discussed in the above "Play attacks" section.
+
```

Comment:
How are replay attacks affected by this? Not at all? 
Are PayJoins still possible to do as before?
Are there any consensus breaking changes required?

Reply:
* Replay attacks seem unaffected.
* Interestingly, it solves the various privacy concerns I have with PayJoins. Using today's workflows, I consider PayJoins a non-starter, as they're terrible for the receiver's privacy. With this RFC, receiver doesn't have to leak any inputs until they choose to broadcast the transaction.
* Nope, consensus layer remains unchanged.

Reply:
Ohhh... you're right, it doesn't become possible to probe for tx outputs... ping @phyro

Reply:
First, congrats on the scheme David :+1: I think I understand the general idea, but have yet to fully grasp it. Yeah, this should make it easy to avoid the need to show the input to the sender. The receiver not having to send back to sender makes it possible to hide receiver's input which makes it more suitable for the receiver's side. It does require the sender to show their inputs and outputs to the receiver though which is not true in today's tx building since a tx could be merged prior to the receiver seeing it. This scheme basically shifts the roles of information knowledge hence making the sender show their inputs/outputs to the receiver. This gives that possible hiding of inputs/outputs power that the sender has in the current transaction building process to the receiver if I understand it correctly.

As you've already said, repetitive utxo spoofing should be impossible in theory, but some care needs to be taken because if the `receive` action is automated, the sender could send 1 nanogrin a dozen times and then observe the transactions which could cause output leaking if the receiver finalizing the tx was automated.

Such `A->B` scheme would be nice if we conclude it is secure. I have some questions but am not sure how relevant they are:
1. What happens if the sender sends to the same address twice? are there any possible security issues?
2. UTXO spoofing should be impossible in theory, but what happens if you just send 1 nanogrin a dozen times to a wallet that is automated to receive? My current understanding is that automated PayJoin `receive` transactions could be a bad idea if one is concerned about this attack, so payjoins might be best if they required a manual confirmation. Am I understanding it correctly?

Reply:
> This scheme basically shifts the roles of information knowledge hence making the sender show their inputs/outputs to the receiver.

Yea, this is how we used to do it before compact slates anyway. It's far less harmful than letting a random sender prod a listening wallet for inputs. The sender chooses who to reveal its inputs to, making it far less worrisome.



> the sender could send 1 nanogrin a dozen times and then observe the transactions which could cause output leaking if the receiver finalizing the tx was automated.

You're talking about when using PayJoin? 



> 1. What happens if the sender sends to the same address twice? are there any possible security issues?

Huge security issues, which is why I define the exact process in the "Nonce Reuse" section. Have a re-read of it and let me know if you still have questions.



> 2\. My current understanding is that automated PayJoin `receive` transactions could be a bad idea if one is concerned about this attack, so payjoins might be best if they required a manual confirmation. Am I understanding it correctly?

Yea, you're understanding that correctly. Good point. PayJoin is still problematic. However, since these are one-time addresses, presumably you're not giving dozens of them out to the same person. You *could* have an automated service that does that though, so it's still worth considering if maybe PayJoin should always require manual intervention. An automated service could still collect partial transactions, and user could later manually finalize them all and choose whether to payjoin or not.